File: adapters/archidoc-rust/src/walker.rs
use std::fs;
use std::path::Path;

use archidoc_types::ModuleDoc;
use walkdir::WalkDir;

use crate::parser;
use crate::path_resolver;

/// Walk a source tree and extract ModuleDocs from all module entry files.
///
/// Finds `lib.rs` and `mod.rs` files, extracts `//!` doc comments,
/// and builds ModuleDoc structs from the parsed annotations.
pub fn extract_all_docs(root: &Path) -> Vec<ModuleDoc> {
    let mut docs = Vec::new();

    for entry in WalkDir::new(root).into_iter().filter_map(|e| e.ok()) {
        let path = entry.path();

        let filename = match path.file_name().and_then(|n| n.to_str()) {
            Some(name) if name == "lib.rs" || name == "mod.rs" => name,
            _ => continue,
        };

        let content = match parser::archidoc_from_file(path) {
            Some(c) if !c.trim().is_empty() => c,
            _ => continue,
        };

        let module_path = path_resolver::path_to_module_name(path, root, filename);
        let c4_level = parser::extract_c4_level(&content);
        let pattern = parser::extract_pattern(&content);
        let pattern_status = parser::extract_pattern_status(&content);
        let description = parser::extract_description(&content);
        let parent_container = parser::extract_parent_container(&module_path);
        let relationships = parser::extract_relationships(&content);
        let files = parser::extract_file_table(&content);

        docs.push(ModuleDoc {
            module_path,
            content,
            source_file: path.to_string_lossy().to_string(),
            c4_level,
            pattern,
            pattern_status,
            description,
            parent_container,
            relationships,
            files,
        });
    }

    docs.sort_by(|a, b| a.module_path.cmp(&b.module_path));
    docs
}

/// Read all `.rs` source files in a directory and return their contents.
///
/// Returns a vec of `(filename, source_code)` pairs. Skips files that
/// cannot be read. Used by pattern heuristics to scan a module directory
/// for structural evidence without coupling AST analysis to filesystem I/O.
pub fn read_rs_sources(dir: &Path) -> Vec<(String, String)> {
    let entries = match fs::read_dir(dir) {
        Ok(e) => e,
        Err(_) => return Vec::new(),
    };

    entries
        .filter_map(|e| e.ok())
        .filter_map(|entry| {
            let path = entry.path();
            if path.extension().and_then(|e| e.to_str()) == Some("rs") {
                let filename = path.file_name()?.to_string_lossy().to_string();
                let source = fs::read_to_string(&path).ok()?;
                Some((filename, source))
            } else {
                None
            }
        })
        .collect()
}

File: adapters/archidoc-md/src/walker.ts
/**
 * @c4 component
 *
 * Walks a directory tree finding _index.md annotation files.
 *
 * | File | Pattern | Purpose | Health |
 * |------|---------|---------|--------|
 * | `walker.ts` | -- | Directory traversal for _index.md files | stable |
 */

import * as fs from "node:fs";
import * as path from "node:path";

import * as parser from "./parser.js";
import { pathToModuleName } from "./path-resolver.js";
import type { ModuleDoc } from "./types.js";

const SKIP_DIRS = new Set([
  "node_modules",
  ".git",
  "target",
  "dist",
  ".ragd",
  ".vite",
  ".claude",
]);

/**
 * Walk a documentation tree and extract ModuleDocs from all _index.md files.
 *
 * Finds `_index.md` files, extracts HTML comment @c4 annotations, and builds
 * ModuleDoc structs from the parsed content.
 *
 * The root-level _index.md (if present) is skipped — it has no meaningful
 * module_path and describes the repo itself, not a sub-module.
 */
export function extractAllDocs(root: string): ModuleDoc[] {
  const docs: ModuleDoc[] = [];
  root = path.resolve(root);

  walkDir(root, (filePath) => {
    if (path.basename(filePath) !== "_index.md") return;

    let content: string;
    try {
      content = fs.readFileSync(filePath, "utf-8");
    } catch {
      return;
    }

    const modulePath = pathToModuleName(filePath, root);
    if (!modulePath) return;

    const c4Level = parser.extractC4Level(content);
    const description = parser.extractDescription(content);
    const parentContainer = parser.extractParentContainer(modulePath);
    const relationships = parser.extractRelationships(content);
    const files = parser.extractFileTable(content);

    docs.push({
      module_path: modulePath,
      content,
      source_file: filePath,
      c4_level: c4Level,
      pattern: "--",
      pattern_status: "planned",
      description,
      parent_container: parentContainer,
      relationships,
      files,
    });
  });

  docs.sort((a, b) => a.module_path.localeCompare(b.module_path));
  return docs;
}

function walkDir(dir: string, callback: (filePath: string) => void): void {
  let entries: fs.Dirent[];
  try {
    entries = fs.readdirSync(dir, { withFileTypes: true });
  } catch {
    return;
  }

  for (const entry of entries) {
    if (SKIP_DIRS.has(entry.name)) continue;

    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      walkDir(fullPath, callback);
    } else if (entry.isFile()) {
      callback(fullPath);
    }
  }
}
